### ISSUE TITLE
Docker Swarm Deploy example - multi master

### ISSUE THREAD

**Issue by csib** [bitnami/bitnami-docker-mariadb-galera#13]
Hi,

Please extend/add docker-compose.yml file for a multi-master replication cluster in Swarm. I've done it on my own, but I am not an expert like you, and my log is full with:

> 2020-03-25 16:59:36 0 [Note] WSREP: (07f45356, 'tcp://0.0.0.0:4567') reconnecting to ef87323b (tcp://10.0.1.21:4567), attempt 121950

I think this is not a healthy behaviour, but everything seems functional.

Thank you!

**Comment by juan131** [comment]
Hi @csib 

Thanks so much for your suggestion.

At this very moment, our pipeline tests the containers locally (using "docker-compose") and on different K8s distributions (using the associated Helm chart). However, **we do not test our containers using Docker Swarm.**

We want to ensure that whatever we provide through this repo is tested on a daily basis. Therefore, we cannot provide a "docker-compose.yaml" for Swarm since we don't have the tools in place to test it properly, and we cannot guarantee is working as expected.

**Comment by Ornias1993** [comment]
@csib To add a bit: It isn't that hard, but you might need a bit of compose-fu.
You need to check how you handle hostnames. If you like you can post your compose.yml here and have people take a look. if everything is alright

However I also understand @juan131 very well:
The moment you supply example config (even when you add a big NOT-SUPPORTED on top), people expect it to be supported.

**Comment by Ornias1993** [comment]
@csib Lets be helpfull.. I set one up for you:

```
version: '3'
services:
  galera0:
      image: bitnami/mariadb-galera:latest
      hostname: galera0
      networks:
        - internal
      environment:
        - MARIADB_GALERA_CLUSTER_BOOTSTRAP=yes
        - MARIADB_ROOT_PASSWORD=PASSWORD
        - MARIADB_GALERA_MARIABACKUP_PASSWORD=PASSWORD

      volumes:
        - /var/data/mariadb/galera0/db:/bitnami/mariadb
        - /var/data/mariadb/galera0/config:/opt/bitnami/mariadb/conf/

  galera:
      image: bitnami/mariadb-galera:latest
      hostname: galera{{.Task.Slot}}
      networks:
        - internal
      environment:
         - MARIADB_GALERA_CLUSTER_ADDRESS=gcomm://galera0
         - MARIADB_ROOT_PASSWORD=PASSWORD
         - MARIADB_GALERA_MARIABACKUP_PASSWORD=PASSWORD
      volumes:
        - /var/data/mariadb/galera{{.Task.Slot}}/db:/bitnami/mariadb
        - /var/data/mariadb/galera{{.Task.Slot}}/config:/opt/bitnami/mariadb/conf/



#  maxscale:
#      image: 'mariadb/maxscale:latest'
#      networks:
#          - internal
#      depends_on:
#          - master
#          - slave
#      volumes:
#          - /var/data/mariadb/maxscale:/etc/maxscale.cnf.d
#      ports:
#          - "4006:4006"
#          - "4008:4008"
#          - "8989:8989"
networks:
  internal:
    driver: overlay
    ipam:
      config:
        - subnet: 172.16.70.0/24

```

Note: Also added example maxscale config, commented out (because it isn't plug and play).
One thing I'm not happy about and didn't fully figure out is how to make sure the permissions are right. Installing required manual intervention a few times running `chown -R 1001:1001 /var/data/mariadb` in the background

**Comment by juan131** [comment]
Hi @Ornias1993 

Thanks so much for sharing this solution for Docker Swarm!! We really appreciate your collaboration.

 > The moment you supply example config (even when you add a big NOT-SUPPORTED on top), people expect it to be supported.

+100 
That's something we want to avoid, even though we'll reach a wider set of users providing them. However, if we don't have a way to automatically test it on a daily basis, then we prefer not to publish it.

**Comment by Ornias1993** [comment]
> Thanks so much for sharing this solution for Docker Swarm!! We really appreciate your collaboration.

The permission issues and such do prove your point though: it is not just a copy-and-paste and does need some custom TLC. Afaik the issue is the difference in how resp. Docker Swarm and k8s, handle container permissions. But someone with more experience in docker permissions should take a look at that.

> +100
> That's something we want to avoid, even though we'll reach a wider set of users providing them. However, if we don't have a way to automatically test it on a daily basis, then we prefer not to publish it.

I might take a look at it someday, afaik mariadb has their own test code public using docker swarm.


Another thing I want to point out:
`- MARIADB_GALERA_CLUSTER_BOOTSTRAP=yes` requires an actual "yes", it would be nice if we can also accept "1". 
`- MARIADB_GALERA_CLUSTER_ADDRESS=gcomm://galera0` when used on a bootstrap node goes haywire (as expected), would it be too much to ask to just overwrite to default when `MARIADB_GALERA_CLUSTER_BOOTSTRAP` is set?

If those two fixes are included we could do this:

```
  galera:
      image: bitnami/mariadb-galera:latest
      hostname: galera{{.Task.Slot}}
      networks:
        - internal
      environment:
        - MARIADB_GALERA_CLUSTER_BOOTSTRAP={{.Task.Slot}}
         - MARIADB_GALERA_CLUSTER_ADDRESS=gcomm://galera1
         - MARIADB_ROOT_PASSWORD=PASSWORD
         - MARIADB_GALERA_MARIABACKUP_PASSWORD=PASSWORD

      volumes:
        - /var/data/mariadb/galera{{.Task.Slot}}/db:/bitnami/mariadb
        - /var/data/mariadb/galera{{.Task.Slot}}/config:/opt/bitnami/mariadb/conf/
```

Which means with swarm we can let swarm handle balancing all the node and make sure both the bootstrap and normal nodes are spread out.

With seperated bootstrap and master services, this might lead to both bootstrap and 1 of the masters being put on the same node as the bootstrap. If that node fails everything goes down with it.

**Comment by juan131** [comment]
Hi @Ornias1993 

> - `MARIADB_GALERA_CLUSTER_BOOTSTRAP=yes` requires an actual `yes`, it would be nice if we can also accept `1`.

We found issues in the past with `true/false`, because some expressions such as `if [[ "$FOO" = true]]` where problematic since some shells consider the expression "true" when the environment variable exists (no matter the value it has). That's why we started to use "yes/no" so we could do strings comparisons. 

That said, accepting `0/1` shouldn't be problematic. We just need to adapt the code to accept new values. My question is: why is that important? I mean it's just a convention and it's very easy to translate from `0/1` to `true/false` or `yes/no`. See some examples:

```bash
[[ $FOO -eq 0 ]] && BAR="yes" || BAR="no"
[[ $FOO -eq 0 ]] && BAR=true || BAR=false
```

> - MARIADB_GALERA_CLUSTER_ADDRESS=gcomm://galera0 when used on a bootstrap node goes haywire (as expected), would it be too much to ask to just overwrite to default when

Why don't you omit the environment variable `MARIADB_GALERA_CLUSTER_BOOTSTRAP` and let the nodes decide if they are the "boostrap node" based on their address? See the function below:

https://github.com/bitnami/bitnami-docker-mariadb-galera/blob/master/10.4/debian-10/rootfs/opt/bitnami/scripts/libmysqlgalera.sh#L29

**Comment by Ornias1993** [comment]
@juan131 why I didnt do that is quite simple: docker compose/stack is a picky beast. It doesnt seem to accept some forms of if-statements and variables in some places.

Could try though...

About that auto bootstrat function: why isnt this documented? I didnt use it because it isnt documented. How should I know it existed?

**Comment by juan131** [comment]
> why I didn't do that is quite simple: docker compose/stack is a picky beast. It doesn't seem to accept some forms of if-statements and variables in some places.

But it accepts strings, doesn't it? I mean, you should be able to set "yes/no" with no major issues, right?

> About that auto bootstrat function: why isnt this documented? I didnt use it because it isnt documented. How should I know it existed?

I agree with you, we should mention in [this section](https://github.com/bitnami/bitnami-docker-mariadb-galera#setting-up-a-multi-master-cluster) that the bootstrap node is calculated based on the `MARIADB_GALERA_CLUSTER_ADDRESS` when the `MARIADB_GALERA_CLUSTER_BOOTSTRAP` environment variable is not set.

**Comment by Choogster1** [comment]
I'm also having a problem with getting Galera working on Swarm correctly. I'm trying to connect Moodle to Galera, but Moodle refuses to recognise that there is a database server running (works fine with Mariadb 10.4). I have  mariadb-galera-0 and mariadb-galera created along with moodle all running behind Traefik reverse proxy. If Moodle can connect to Mariadb with no problems at all, how come it doesn't even find Galera? There's no connection errors such as wrong password, it just won't even find the server...  if anybody can shed some light it would be appreciated.

**Comment by juan131** [comment]
Hi @Choogster1 

Could you please share the error Moodle is throwing when trying to connect Galera? 
Does Moodle complains about not being able to reach the Galera server, or is it complaining about a specific database not existing in that server?

**Comment by Choogster1** [comment]
> Hi @Choogster1
> 
> Could you please share the error Moodle is throwing when trying to connect Galera?
> Does Moodle complains about not being able to reach the Galera server, or is it complaining about a specific database not existing in that server?

Hi Juan.

The error Moodle throws is purely 'unable to connect to database server' so it can't even find the database to try to connect to.

**Comment by juan131** [comment]
Hi @Choogster1 

Could you please share the Docker Compose you're using to deploy MariaDB Galera and Moodle? I'd like to reproduce the same environment on my side so I can see what's going on.

**Comment by Choogster1** [comment]
> Hi @Choogster1
> 
> Could you please share the Docker Compose you're using to deploy MariaDB Galera and Moodle? I'd like to reproduce the same environment on my side so I can see what's going on.

Turns out it looks like it's a problem with Azure, running Galera and Moodle on other non-cloud servers worked fine. Probably a security or firewall setting.

**Comment by juan131** [comment]
Hi @Choogster1 

> Probably a security or firewall setting.

It's very likely that you have to configure a Network Security Group to allow connections between Moodle and MariaDB Galera, it makes sense.

Thanks for sharing with the rest what you found out. I'm very sure it'll help other users.

**Comment by miberecz** [comment]
> Hi @Ornias1993
> 
> > * `MARIADB_GALERA_CLUSTER_BOOTSTRAP=yes` requires an actual `yes`, it would be nice if we can also accept `1`.
> 
> We found issues in the past with `true/false`, because some expressions such as `if [[ "$FOO" = true]]` where problematic since some shells consider the expression "true" when the environment variable exists (no matter the value it has). That's why we started to use "yes/no" so we could do strings comparisons.
> 
> That said, accepting `0/1` shouldn't be problematic. We just need to adapt the code to accept new values. My question is: why is that important? I mean it's just a convention and it's very easy to translate from `0/1` to `true/false` or `yes/no`. See some examples:
> 
> ```shell
> [[ $FOO -eq 0 ]] && BAR="yes" || BAR="no"
> [[ $FOO -eq 0 ]] && BAR=true || BAR=false
> ```
> 
> > * MARIADB_GALERA_CLUSTER_ADDRESS=gcomm://galera0 when used on a bootstrap node goes haywire (as expected), would it be too much to ask to just overwrite to default when
> 
> Why don't you omit the environment variable `MARIADB_GALERA_CLUSTER_BOOTSTRAP` and let the nodes decide if they are the "boostrap node" based on their address? See the function below:
> 
> https://github.com/bitnami/bitnami-docker-mariadb-galera/blob/master/10.4/debian-10/rootfs/opt/bitnami/scripts/libmysqlgalera.sh#L29

Sorry for bumping an old topic, but would you elaborate on this "nodes decide if they are the boostrap node"
The link to the code is broken now. 
I have tried to omit this settings from a swarm service with 3 replica, and all three created their on galera-cluster separately from the others.

**Comment by javsalgar** [comment]
Hi,

By default the nodes will check their status file and will know if they can bootstrap the cluster or not. The environment variable mentioned there is for forcing the bootstrap in a specific node.

**Comment by jerrac** [comment]
Yes, one location for the data is not my favorite thing in the world. But with our current resources, there's no way around it. My goal for now it to set things up so that when we can afford to do it right, the transition is easy. The joys of working for a small community college.

As for your suggestion, it didn't work.

Initial attempts were via Portainer on my Docker Swarm cluster, I also tried straight from a manager node with `docker stack deploy` and even just a `docker compose up`. Oh, and I tried putting quotes around the `device:` value, and not putting quotes around it.

I also tried creating the volume on the command line, then deploying the stack or service. Didn't work.

I'm guessing that the template syntax is specific to the service command, so it just won't work in the context of volumes.

I think I'll see if I can do it with local storage, it's all the same SAN underneath anyway... :(
